src/main/java/bio/terra/workspace/service/workspace/flight/DeleteWorkspaceAuthzStep.java

```diff
@@ -0,0 +1,40 @@
+package bio.terra.workspace.service.workspace.flight;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamService;
+import java.util.UUID;
+
+public class DeleteWorkspaceAuthzStep implements Step {
+
+  private SamService samService;
+  private AuthenticatedUserRequest userReq;
+
+  public DeleteWorkspaceAuthzStep(SamService samService, AuthenticatedUserRequest userReq) {
+    this.samService = samService;
+    this.userReq = userReq;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext) throws RetryException {
+    FlightMap inputMap = flightContext.getInputParameters();
+    UUID workspaceID = inputMap.get(WorkspaceFlightMapKeys.WORKSPACE_ID, UUID.class);
+    samService.deleteWorkspace(userReq.getRequiredToken(), workspaceID);
+    return StepResult.getStepResultSuccess();
+  }
+
+  // Note: This doesn't fully undo the do-step, as the newly-created IAM resource will not have
+  // the same access policies as the old one. However, it prevents strange half-deleted states
+  // from occuring.
+  @Override
+  public StepResult undoStep(FlightContext flightContext) {
+    FlightMap inputMap = flightContext.getInputParameters();
+    UUID workspaceID = inputMap.get(WorkspaceFlightMapKeys.WORKSPACE_ID, UUID.class);
+    samService.createWorkspaceWithDefaults(userReq.getRequiredToken(), workspaceID);
```

Comment:
I don't think that this will work. Sam prevents ID reuse for workspaces, so we won't actually be able to re-create the workspace resource in Sam if this fails.

I am trying to think of another approach for dealing with this failure but I haven't landed on anything yet. If the workspace resource is deleted, then it will appear deleted for the user even if it's still in the WSM db... but that doesn't mean that all of the contents of the workspace were deleted- so users could falsely be led to believe that they're not going to be accruing costs associated with the workspace anymore. It's a tricky case to handle. It adds a lot of complexity but maybe we could periodically sweep for orphaned parts of the workspace state?

Does stairway allow for us to retry? We could put our best effort in to retry the step until it's eventually deleted and log it if we can't delete it.

A lot of this was just me thinking out loud...

Reply:
Good to know on the ID reuse. It already felt a little weird to try re-creating a deleted resource, but if we can't do that at all it seems better to just do the same thing as the `DeleteWorkspaceStateStep` and tell Stairway there's nothing we can do.
I added a retry on Sam errors in the `do` step which should handle transient errors, but I think ultimately if we hit the undo step the best we can do is log this as a dismal failure.
That sweep might make sense at some point, but for now we can probably log/monitor dismal failures from Stairway and take manual action from there.